File: octo/mns/detector/src/main/java/com/sankuai/octo/util/ScanUtils.java
package com.sankuai.octo.util;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONObject;
import com.sankuai.inf.octo.mns.ProcessInfoUtil;
import com.sankuai.octo.Common;
import com.sankuai.octo.model.Provider;
import com.sankuai.octo.model.report.ScannerReport;
import org.apache.commons.lang.StringUtils;
import org.apache.zookeeper.KeeperException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.UnsupportedEncodingException;
import java.net.InetAddress;


public class ScanUtils {
    private final static Logger logger = LoggerFactory.getLogger(ScanUtils.class);
    public static String hostIpPrefix = "";

    static {
        if (StringUtils.isBlank(hostIpPrefix))
            hostIpPrefix = getLocalIpV4Prefix();
    }

    public static String getLocalIpV4Prefix() {
        String localAddress = ProcessInfoUtil.getLocalIpV4();
        String[] ipSegs = localAddress.split("\\.");
        String prefix = ipSegs[0] + "." + ipSegs[1] + ".";
        return prefix;

    }

    public static JSONObject bytes2JO(byte[] bytes) {
        if (null == bytes)
            return null;
        String serverInfo = "";
        JSONObject jo = null;
        try {
            serverInfo = new String(bytes, "utf-8");
            jo = JSON.parseObject(serverInfo);
        } catch (UnsupportedEncodingException e) {
            logger.error(e.getMessage(), e);
        } catch (Exception e) {
            logger.error(e.getMessage(), e);
        } finally {
            return jo;
        }
    }

    public static boolean isReachable(String ip, int timeout) {
        boolean result = true;
        try {
            InetAddress inetAddress = InetAddress.getByName(ip);
            result = inetAddress.isReachable(timeout);
        } catch (Exception e) {
            result = false;
        } finally {
            return result;
        }
    }


    public static JSONObject getProviderJsonInZK(String serverPath) {
        return ScanUtils.bytes2JO(Common.zkClient.get(serverPath));
    }

    public static void deleteUnReachableProvider(Provider provider) {
        for (int i = 0; i < 3; i++) {
            if (isReachable(provider.getIp(), Common.getTimeOutInMills(provider.getIp())))
                return;
        }

        logger.warn("Delete unReachable provider! " + provider.getIdentifierString());
        try {
            Common.zkClient.delete(provider.getServerPath());
            updateClusterTime(provider.getProvidersDir());
        } catch (KeeperException e) {
            logger.warn("deleteUnReachableProvider Exception" + e.getCause());
        } finally {
            SendReport.send(new ScannerReport(1, "DeleteProvider",
                    "Delete unReachable provider!", provider.getIdentifierString()));
        }
    }


    public static void updateClusterTime(String providersDir) {
        JSONObject clusterJO = ScanUtils.bytes2JO(Common.zkClient.get(providersDir));
        long lastUpdateTime = System.currentTimeMillis() / 1000;
        clusterJO.put("lastUpdateTime", lastUpdateTime);
        Common.zkClient.update(providersDir, JSON.toJSONBytes(clusterJO), false);
        logger.info("cluster status changed " + providersDir + " lastUpdateTime:" + lastUpdateTime);

    }

}

File: octo/mns/updater/src/main/scala/com/sankuai/octo/updater/util/ScanUtils.scala
package com.sankuai.octo.updater.util

import java.nio.charset.StandardCharsets.UTF_8

import com.alibaba.fastjson.{JSON, JSONObject}
import com.sankuai.meituan.config.listener.IConfigChangeListener
import com.sankuai.octo.scanner.model.Provider
import com.sankuai.octo.scanner.model.report.UpdateStatusReport
import com.sankuai.octo.util.SendReport
import org.slf4j.LoggerFactory

/**
  * Created by jiguang on 14-9-29.
  */
object ScanUtils {
  private val logger = LoggerFactory.getLogger(this.getClass)

  private val UPDATER_SWITCH = "updaterSwitch"

  @volatile
  private var switch = config.get(UPDATER_SWITCH, "false").toBoolean

  config.addListener(UPDATER_SWITCH, new IConfigChangeListener {
    override def changed(key: String, oldValue: String, newValue: String): Unit = {
      logger.info(s"key:$key,oldValue:$oldValue,newValue:$newValue")
      switch = newValue.toBoolean
    }
  })


  def getProviderJsonInZK(serverPath: String): JSONObject = {
    ScanUtils.bytes2JO(Common.zkClient.get(serverPath))
  }

  def updateServerStatus(validatedStatus: Int, status: Int, provider: Provider, jo: JSONObject) {
    val strLog = "update node status:" + provider.getServerPath + " status, from " + status + " to " + validatedStatus
    logger.info(strLog)
    jo.put("status", validatedStatus)
    jo.put("lastUpdateTime", System.currentTimeMillis / 1000)
    if (switch) {
      Common.zkClient.update(provider.getServerPath, JSON.toJSONBytes(jo), false)
    }
    updateClusterTime(provider.getProvidersDir)
    SendReport.send(new UpdateStatusReport(0, "UpdateStatus", strLog, status, validatedStatus, provider.getIdentifierString))
  }

  def updateClusterTime(providersDir: String) {
    val clusterJO: JSONObject = ScanUtils.bytes2JO(Common.zkClient.get(providersDir))
    val lastUpdateTime: Long = System.currentTimeMillis / 1000
    clusterJO.put("lastUpdateTime", lastUpdateTime)
    if (switch) {
      Common.zkClient.update(providersDir, JSON.toJSONBytes(clusterJO), false)
    }
  }

  def bytes2JO(bytes: Array[Byte]): JSONObject = {
    if (null == bytes) {
      return null
    }
    val serverInfo = new String(bytes, UTF_8)
    try {
      val jo = JSON.parseObject(serverInfo)
      jo
    } catch {
      case e: Exception =>
        logger.error("Exception ", e)
        null
    }

  }
}
